modules/ml/src/main/java/org/apache/ignite/ml/dataset/UpstreamTransformerChain.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.dataset;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Stream;
+import org.apache.ignite.ml.math.functions.IgniteFunction;
+
+/**
+ * Class representing chain of transformers applied to upstream.
+ *
+ * @param <K> Type of upstream keys.
+ * @param <V> Type of upstream values.
+ */
+public class UpstreamTransformerChain<K, V> implements Serializable {
+    /** Seed used for transformations. */
+    private Long seed;
+
+    /** List of upstream transformations. */
+    private List<UpstreamTransformer<K, V>> list;
```

Comment:
Can we reduce these transformers using `andThen` approach?

Reply:
It can be reduced in theory, but it would be a bit cumbersome since each upstream transformer is a bi-function. I think this way it is more readable.

Reply:
And again, it's bi-function only because we don't solve design issue with seeds. Let's solve it and it will lead to significant simplification in this part as well.